### ISSUE TITLE
Replace `collections.namedtuple` with `dataclasses` or `attrs`

### ISSUE THREAD

**Issue by dutc** [Telecominfraproject/oopt-gnpy#132]
Referencing #100, we need to improve our `Elements` base class to better handle optional fields while still validating required fields.

There are two major options we can choose for this:
- `dataclasses` (built into Python >3.7, available via `pip install dataclasses` prior to that) https://www.python.org/dev/peps/pep-0557/
- `attrs` (user-supported library) https://github.com/python-attrs/attrs

We want to preserve the immutable behavior of network elements for ease-of-reasoning and optimisation purposes. We want to improve the user interface when creating new fields and improve the robustness of our code in validating user input on required fields.

**Comment by ojnas** [comment]
I have a couple of questions related to the "immutable behavior of network elements":

1. Why is not the Edfa element class using an immutable namedtuple for `params` like all the other element classes? This looks a bit inconsistent.

2. What exactly is the importance of this immutable behavior? Currently, for other elements than Edfa, the immutable `params` is anyway not used after initialization, where it is just used for assigning values to other (mutable) variables. For example in `class Roadm` you have `self.loss = self.params.loss`. Then this `roadm.loss` variable is updated by `build_network` in `network.py`. Same for Fiber class variables `con_in` and `con_out` for example.

**Comment by jeanluc-auge** [comment]
I have check the data class: I like it a lot.
Only drawback it needs python 3.7, which is a bit of a pain to update.
